Add DynamicalFriction Test Problem

ej-1211 opened the pull request.

JeiwHuang reviewed (changes requested):
Hi @ej-1211 ,

I am really sorry for the delayed review. 😢 
Thank you for your patience.
I've had a quick look at this PR recently.

I can run the three default cases and reproduce similar results in this [page](https://hackmd.io/@Q-7np4JqRmmhXMvo0h5Fqg/B1L_61Yb6).
However, the figure for the `Quick_Burkert` is not 100% identical with the reference. Here is my results:
![Compare_result_20240201-1040](https://github.com/gamer-project/gamer/assets/24509818/438493b0-1da5-408b-b8db-587a7589a053)
I am not sure why and I didn't figure out what may went wrong.

I provided some suggestions in the inline comments. Please take a look.
Please let me know if my messages are not clear or I misunderstood anything. 🙏  Thanks!

hyschive commented on src/TestProblem/Hydro/DynamicalFriction/Init_TestProb_Hydro_DynamicalFriction.cpp:
```diff
@@ -0,0 +1,347 @@
+	}
+	fprintf(File, "%13.7f\t%13.7e\t%13.7e\t%13.7e\n",
+           Time[0]  ,amr->BoxCenter[0], amr->BoxCenter[1], amr->BoxCenter[2]);
+	fclose ( File );	
```
Please remove _all_ tabs instead of just trailing tabs. Tabs should only be used in the makefile.

ej-1211 replied:
done

hyschive commented on src/TestProblem/Hydro/DynamicalFriction/Init_TestProb_Hydro_DynamicalFriction.cpp:
```diff
@@ -0,0 +1,347 @@
+
+if ( MPI_Rank == 0)
+{
+	char Filename[MAX_STRING];
+	sprintf( Filename, "%s", "Record__RESULT_Center.txt" );
+	FILE *File = fopen( Filename, "a" );
+        if (Time[0]==0.0){
+        fprintf(File, "%15s%15s%15s%15s\n", "Time", "CenterX", "CenterY", "CenterZ");
+	}
+	fprintf(File, "%13.7f\t%13.7e\t%13.7e\t%13.7e\n",
+           Time[0]  ,Extrema.Coord[0], Extrema.Coord[1], Extrema.Coord[2]);
+	fclose ( File );
+	
+	char Filename_[MAX_STRING];
+	sprintf( Filename_, "%s", "Record__Previous_Center.txt" );
+	FILE *File_ = fopen( Filename_, "a" );
+        if (Time[0]==0.0){
+        fprintf(File_, "%15s%15s%15s%15s%15s\n", "Time", "CenterX", "CenterY", "CenterZ","Search R");
+	}
+	fprintf(File_, "%13.7f\t%13.7e\t%13.7e\t%13.7e\t%13.7e\n",
+           Time[0]  ,min_pot_last[0], min_pot_last[1], min_pot_last[2],Extrema.Radius);
+	fclose ( File_ );
```
Agree. It's because different people may adopt different tab sizes, which breaks code alignment.

ej-1211 replied:
done

hyschive commented:
@JeiwHuang Thanks for the extremely careful and thorough review!

@ej-1211: I agree with @JeiwHuang's most comments. For example,
- Simplify the initial condition construction, if possible, by utilizing the built-in tools in gamer (e.g., `Aux_ComputeProfile()`).
- Make variable names more descriptive
- Add comments whenever appropriate
- Follow the gamer coding format (e.g., no tabs, three-space indent, code alignment)
- Minimize redundant code lines

If you have any questions or different opinions, please feel free to speak out 😉

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Init_TestProb_Hydro_DynamicalFriction.cpp:
- Changed `Par_Init_ByFunction_GC` to `Par_Init_ByFunction_DynamicalFriction`
- Changed `ExtPot_GC.cpp` to `ExtPot_DynamicalFriction.cpp`
- Changed `ExtPot_GC.cu` to `ExtPot_DynamicalFriction.cu`
- Changed `ExtPot_Cu` to `ExtPot_DynamicalFriction`
- Changed `SetExtPotAuxArray_GC` to `SetExtPotAuxArray_DynamicalFriction`
- Changed `Par_Init_ByFunction_GC.cpp` to `Par_Init_ByFunction_DynamicalFriction.cpp`
- Changed `Par_Init_ByFunction_GC` to `Par_Init_ByFunction_DynamicalFriction`
- Changed `Aux_Record_User_GC` to `Aux_Record_User_DynamicalFriction`
- **some variables in the function still need to be fixed**

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/global_var.h:
Declared all the variables in `Par_Init_ByFunction_Dynamical_Friction.cpp`

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Init_TestProb_Hydro_DynamicalFriction.cpp:
```diff
@@ -0,0 +1,347 @@
+if ( MPI_Rank == 0)
+{
+	char Filename[MAX_STRING];
+	sprintf( Filename, "%s", "Record__RESULT_Center.txt" );
```
Done

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Init_TestProb_Hydro_DynamicalFriction.cpp:
```diff
@@ -0,0 +1,347 @@
+           Time[0]  ,Extrema.Coord[0], Extrema.Coord[1], Extrema.Coord[2]);
+	fclose ( File );
+	
+	char Filename_[MAX_STRING];
+	sprintf( Filename_, "%s", "Record__Previous_Center.txt" );
+	FILE *File_ = fopen( Filename_, "a" );
+        if (Time[0]==0.0){
+        fprintf(File_, "%15s%15s%15s%15s%15s\n", "Time", "CenterX", "CenterY", "CenterZ","Search R");
+	}
+	fprintf(File_, "%13.7f\t%13.7e\t%13.7e\t%13.7e\t%13.7e\n",
+           Time[0]  ,min_pot_last[0], min_pot_last[1], min_pot_last[2],Extrema.Radius);
+	fclose ( File_ );
```
Deleted the output of the previous center

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Init_TestProb_Hydro_DynamicalFriction.cpp:
```diff
@@ -0,0 +1,347 @@
+	sprintf( Filename, "%s", "Record__RESULT_Center.txt" );
+	FILE *File = fopen( Filename, "a" );
+        if (Time[0]==0.0){
+        fprintf(File, "%15s%15s%15s%15s\n", "Time", "CenterX", "CenterY", "CenterZ");
+	}
+	fprintf(File, "%13.7f\t%13.7e\t%13.7e\t%13.7e\n",
```
Added # and have aligned with the header

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Init_TestProb_Hydro_DynamicalFriction.cpp:
```diff
@@ -0,0 +1,347 @@
+// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+// !!!!! This routine assume there is only one GC particle !!!!!
+// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+double GC_x=99999.9;
+double GC_y=99999.9;
+double GC_z=99999.9;
```
changed to `double GC_x, GC_y, GC_z; `

ej-1211 commented:
- Followed the **1st suggestion** above (working but not in GAMER coding style yet)
- `Aux_Record_User_DynamicalFriction()` in `Init_TestProb_Hydro_DynamicalFriction.cpp` is done
   - Remove unnesassary `Record__RESULT_Previous_Center.txt`
   - changed `RESULT` to `Result`
   - remove `.txt` from the output file name
   - GAMER coding style done 
      - three space indenting 
      - `{ }` in `if else` statement
   - Aligned the Record__Result_* header and added `#`
   - changed `FixCenter` to boolean and turn to `if ( FixCenter )`

ej-1211 commented:
- Separate `Input__TestProb` to `Input__TestProb` and `Input__Profile_Params`
    - `Input__Profile_Params` is for ParticleEquilibriumIC Test problem
    - reduced the amount of `unregcognize parameter`
- Update the input file in `/example`
    - Only save one set of `Input__*`
- [ ]  Add a feature that can automatically update `Cloud_BulkVel*`  in `Input__Profile_Params`, which is calculated in the GAMER routine

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Tools/Density_Profile_Barry_table_for_GAMER.txt:
Removed every `.txt`

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Par_Init_ByFunction_GC.cpp:
```diff
@@ -0,0 +1,189 @@
+       ParData_AllRank[PAR_VELZ][Par_Idx0] = GC_VELZ;
+       ParData_AllRank[PAR_TYPE][Par_Idx0] = PTYPE_GC;
+       // Broadcast the GC information from the root rank
+       GC_xx = GC_POSX;
+       GC_yy = GC_POSY;
+       GC_zz = GC_POSZ;
```
Deleted these useless variables.

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Init_TestProb_Hydro_DynamicalFriction.cpp:
```diff
@@ -0,0 +1,347 @@
+{
+
+// set the center finding method base on Input__TestProb
+if (FixCenter == 1.0){
```
Done

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Init_TestProb_Hydro_DynamicalFriction.cpp:
```diff
@@ -0,0 +1,347 @@
+{
+
+   fluid[DENS] = GC_SmallGas;
+   fluid[MOMX] = 0;
+   fluid[MOMY] = 0;
+   fluid[MOMZ] = 0;
```
changed

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Tools/A_SET_IC.py:
```diff
@@ -0,0 +1,382 @@
+        smooth_transition_factor_ = smooth_transition(x*r0,begin_smooth_r, end_smooth_r, GC_ri)
+
+        if model_name == "Burkert":
+            # return 1e9
+            return 4*np.pi*x*x*(r0**3) * smooth_transition_factor_ * Burkert_dens(x)
+        elif model_name == "NFW":
+            return 4*np.pi*x*x*(r0**3) * smooth_transition_factor_* NFW_dens(x)
+        elif model_name == "Plummer":
+            return 4*np.pi*x*x*(r0**3) * smooth_transition_factor_* Plummer_dens(x)
+        elif model_name == "LC":
+            return 4*np.pi*x*x*(r0**3) * smooth_transition_factor_* LC_dens(x)
+        elif model_name == "King":
+            return 4*np.pi*x*x*(r0**3) * smooth_transition_factor_* King_dens(x)
+        else:
+            return 0
+    f = integrate.nquad(massbase, [[0, x]])[0] * rho_s
+
+
+    return f
+
+
+def getrho(mass, r0, rt, model_name):
+    if model_name == "None":
+        return 0
+    rho_0 = 3.0
+    mass_init = clustermass(rho_0, r0, rt, model_name)
+    rho_0 = rho_0*mass/mass_init
+    return rho_0
+
+
+def create_table(table_filename, r, dens, mass):
+    f = open(table_filename, "w")
+    f.write("#r[pc] density[Msun/pc^3] enclosed_mass[Msun]\n")
+
+    for i in range(len(r)):
+        if dens[i] != 0.:
+            a = str(r[i])+" "+str(dens[i])+" "+str(mass[i]) + "\n"
+            f.write(a)
+
+    f.close()
```
Moved and add some comment for each function in `Calculate_IC.cpp`

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Init_TestProb_Hydro_DynamicalFriction.cpp:
```diff
@@ -0,0 +1,347 @@
+   Aux_Message(stdout, "Setparameter(): %7.5f %7.5f %7.5f \n",GC_xx,GC_yy,GC_zz);
+   Aux_Message(stdout, "Setparameter(): %7.1f %7.1f \n",FixCenter,SearchRadius);
+   }
+   delete ReadPara;
```
Done

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Par_Init_ByFunction_GC.cpp:
```diff
@@ -0,0 +1,189 @@
+
+
+// load run-time parameters
+       const char* FileName = "Input__TestProb";
+       ReadPara_t *ReadPara  = new ReadPara_t;
+    // ********************************************************************************************************************************
+    // ReadPara->Add( "KEY_IN_THE_FILE",      &VARIABLE,              DEFAULT,       MIN,              MAX               );
+    // ********************************************************************************************************************************
+
+    // For GC
+       ReadPara->Add( "GC_MASS",                 &GC_MASS,               NoDef_double,  NoMin_double,     NoMax_double      );
+
+       ReadPara->Add( "GC_POSX",                 &GC_POSX,               NoDef_double,  NoMin_double,     NoMax_double      );  
+       ReadPara->Add( "GC_POSY",                 &GC_POSY,               NoDef_double,  NoMin_double,     NoMax_double      );
+       ReadPara->Add( "GC_POSZ",                 &GC_POSZ,               NoDef_double,  NoMin_double,     NoMax_double      );
+
+       ReadPara->Add( "GC_VELX",                 &GC_VELX,               NoDef_double,  NoMin_double,     NoMax_double      );
+       ReadPara->Add( "GC_VELY",                 &GC_VELY,               NoDef_double,  NoMin_double,     NoMax_double      ); 
+       ReadPara->Add( "GC_VELZ",                 &GC_VELZ,               NoDef_double,  NoMin_double,     NoMax_double      );
+
+       ReadPara->Read( FileName );
```
Removed

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Par_Init_ByFunction_GC.cpp:
```diff
@@ -0,0 +1,189 @@
+      Filename_Loader.Read_Filenames( "Input__TestProb" );
+      long Par_Idx0 = 0;
+
+      for (int k=0; k<Filename_Loader.filenames.Cloud_Num; k++) {
+
+//       initialize Par_EquilibriumIC for each cloud // There is one cloud only, we're just using the ParticleEquilibriem Testprob and modify it
+         Par_EquilibriumIC Cloud_Constructor;
+         Cloud_Constructor.Load_Physical_Params( Filename_Loader.filenames, k, NPar_AllRank-1 );
+         Cloud_Constructor.Init();
+
+//      check whether the particle number of each cloud is reasonable
+        if ( (Par_Idx0 + Cloud_Constructor.params.Cloud_Par_Num) > NPar_AllRank-1 ) {
+           Aux_Error( ERROR_INFO, "particle number doesn't match (%ld + %ld = %ld > %ld) !!\n",
+                       Par_Idx0, Cloud_Constructor.params.Cloud_Par_Num, Par_Idx0+Cloud_Constructor.params.Cloud_Par_Num, NPar_AllRank-1 );
+        }
```
Yes, `Filename_Loader.filenames.Cloud_Num` is always `1`. Removed the `for` loop

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Init_TestProb_Hydro_DynamicalFriction.cpp:
```diff
@@ -0,0 +1,347 @@
+{
+//     --> a helper macro PRINT_WARNING is defined in TestProb.h
+   const long   End_Step_Default = __INT_MAX__;
+   const double End_T_Default    =  10;
```
Fixed

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Init_TestProb_Hydro_DynamicalFriction.cpp:
```diff
@@ -0,0 +1,347 @@
+   fluid[MOMX] = 0;
+   fluid[MOMY] = 0;
+   fluid[MOMZ] = 0;
+#  ifdef GRAVITY
+   fluid[ENGY] = GC_SmallGas;
+#  endif
```
Deleted the `if` statement

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Input__TestProb_Quick_Burkert:
```diff
@@ -0,0 +1,58 @@
+# (4) Inherit from ParticleEquilibrium (basically NO CHANGE NEEDED) 
+GC_SmallGas           1e-10                         #*The maximum tolerance gas density for the simulation [1e-10]
+ParEqmIC_Params_Filenames  Input__TestProb          #!The filenames of the cloud info (which wwe want to set the Equilibrium IC)
+ParEqmIC_Cloud_Num    1                             #!total number of cloud
+AddExtPot             0                             #!whether to add an external potential table [0]
+ExtPot_Table_Name     NONE                          #!External potential table filename.
```
We don't want to support external potential in this test problem. Removed the variables from `Input__TestProb`

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/ExtPot_GC.cpp:
```diff
@@ -0,0 +1,182 @@
+//-------------------------------------------------------------------------------------------------------
+void SetExtPotAuxArray_GC( double AuxArray_Flt[], int AuxArray_Int[], const double Time )
+{
+   double GC_Rho0 = 1.0;
+   double GC_R0   = 0.1;
+   double GC_Center[3] ={1.5,1.5,1.5};
+   double GC_ExtPotMFrac = 1.0;
```
`ExtPot_GC.cpp` is deleted

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Tools/A_SET_IC.py:
```diff
@@ -0,0 +1,382 @@
+            
+
+
+            # split r into two parts
+            # r_original = r[r <= begin_smooth_r]
+            # r_smooth = r[r > begin_smooth_r]
+
+
+            # mass_original = np.array([clustermass(rho_halo, halo_rs, r_original[i], halo_type) for i in range(len(r_original))])
+            # mass_smooth = np.array([clustermass_soft(rho_halo, halo_rs, r_smooth[i], halo_type, \
+                                                    #  smooth_transition(r_smooth[i], begin_smooth_r, end_smooth_r, GC_ri)) for i in range(len(r_smooth))])
+            # 
+            # combine the two parts
+            # mass = np.concatenate((mass_original, mass_smooth), axis=0)
+            # print('haha') 
```
`A_SET_IC.py` is deleted

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Tools/A_SET_IC.py:
```diff
@@ -0,0 +1,382 @@
+
+
+    print("A-3 Calculate the initial conditions from table...")
+    # Use the table to caculate the initial conditions
```
`A_SET_IC.py` is deleted

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Tools/A_SET_IC.py:
`D_Compare.py` is changed to `plot_comparison.py`

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Input__TestProb_Stalling_Burkert:
```diff
@@ -0,0 +1,58 @@
+
+# (4) Inherit from ParticleEquilibrium (basically NO CHANGE NEEDED) 
+GC_SmallGas           1e-10                         #*The maximum tolerance gas density for the simulation [1e-10]
+ParEqmIC_Params_Filenames  Input__TestProb          #!The filenames of the cloud info (which wwe want to set the Equilibrium IC)
```
Fixed

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Init_TestProb_Hydro_DynamicalFriction.cpp:
```diff
@@ -0,0 +1,347 @@
+if ( amr->Par->Mass[p] < PTYPE_GC )  continue;
+   if ( amr->Par->Type[p] == PTYPE_GC) {
+      GC_x = amr->Par->PosX[p]; 
+      GC_y = amr->Par->PosY[p]; 
```
`Init_TestProb_Hydro_DynamicalFriction.cpp` trailing space problem is fixed.

ej-1211 commented:
- Change from `/b_min_factor` and `/b_max_factor` to `*b_min_factor` and `*b_max_factor`
- Throw away the datetime in the result file
- read the parameter from the `log` file and the `Record__Note` File
- removed `A_` `C_` `X_`

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Tools/A_SET_IC.py:
```diff
@@ -0,0 +1,382 @@
+import numpy as np
+import scipy.integrate as integrate
+import pandas as pd
+import datetime
```
Fixed

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Tools/X_some_function.py:
changed to `read_parameters.py`

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Tools/A_SET_IC.py:
```diff
@@ -0,0 +1,382 @@
+    f = open(table_filename, "w")
+    f.write("#r[pc] density[Msun/pc^3] enclosed_mass[Msun]\n")
+
+    for i in range(len(r)):
+        if dens[i] != 0.:
+            a = str(r[i])+" "+str(dens[i])+" "+str(mass[i]) + "\n"
+            f.write(a)
```
`A_SET_IC.py` is deleted, `Calculate_IC.cpp` in `GAMER` has fixed this

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Input__Parameter_NFW_interpolation:
Done

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/submit_eureka.job:
done

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/README.md:
done

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Tools/README.md:
done

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Input__TestProb_NFW_interpolation:
Different parameters are listed in README

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Input__TestProb_Quick_Burkert:
```diff
@@ -0,0 +1,58 @@
+# (2) Total Cloud Parameters
+Cloud_CenterX         10E3                          #*central x coordinate in [pc] SHOULD BE SET BY USER [38E3]
+Cloud_CenterY         10E3                          #*central y coordinate in [pc] SHOULD BE SET BY USER [38E3]
+Cloud_CenterZ         10E3                          #*central z coordinate in [pc] SHOULD BE SET BY USER [38E3]
```
done

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Tools/get_center_of_mass.py:
```diff
@@ -0,0 +1,38 @@
+import yt
+
+# Define a range of file numbers
+file_numbers = range(0, 17)  # This assumes your file names follow the pattern Data_000000, Data_000001, ..., Data_000016
```
`get_center_of_mass.py` is deleted

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Tools/get_center_of_mass.py:
```diff
@@ -0,0 +1,38 @@
+
+# Open the file in append mode and write the results
+with open(output_file, 'a') as f:
+    f.write("Data_ID\tCoM_X\tCoM_Y\tCoM_Z\n")
```
`get_center_of_mass.py` is deleted

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Tools/plot_density_profile.py:
```diff
@@ -0,0 +1,144 @@
+print( '\nCommand-line arguments:' )
+print( '-------------------------------------------------------------------' )
+for t in range( len(sys.argv) ):
+   print str(sys.argv[t]),
```
`plot_density_profile.py` is deleted

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Tools/plot_density_profile.py:
```diff
@@ -0,0 +1,144 @@
+for f in files:
+   
+   ds = yt.load(f)
+   time = round(ds.current_time,3)
```
`plot_density_profile.py` is deleted

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Tools/plot_density_profile.py:
```diff
@@ -0,0 +1,144 @@
+    else:
+        return 0
+y_Burkert = density(562.63979048052,0.25,x,"Burkert")*1E6
+y_Plummer = density(37.12790416639654,0.668,x,"Plummer")*1E6
```
`plot_density_profile.py` is deleted

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Tools/plot_trajectory_movie.py:
Have set the area to assign for different case :
```
GC_path = ''
Center_path = ''
Start_Time = 2000
End_Time = 5000
x_limit = (7000,9000)
y_limit = (7000,9000)
```

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Tools/plot_trajectory_movie.py:
```diff
@@ -0,0 +1,79 @@
+Center_df = pd.read_csv(Center_path, sep='\s+', header=0)
+GC_df = pd.read_csv(GC_path, sep='\s+', header=0)
+Center_df = Center_df[(Center_df['Time'] > 2000) & (Center_df['Time'] < 5000)]
+GC_df = GC_df[(GC_df['Time'] > 2000) & (GC_df['Time'] < 5000)]
```
Have set the area to input the start time and end time

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Input__TestProb_NFW_interpolation:
```diff
@@ -0,0 +1,59 @@
+Cloud_R0              5487.99489144317                         #+ Maximum core radius for the cloud : set to be HALO_Rs as default 
+Cloud_BulkVelX        0.0                             #+ Bulk Velocity for the whole cloud
+Cloud_BulkVelY        -0.00985897009841205                             #+ Bulk Velocity for the whole cloud
+Cloud_BulkVelZ        0.0                           #+ Bulk Velocity for the whole cloudn
```
Done

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Tools/C_Chandrasekhar_dynamical_friction.py:
```diff
@@ -0,0 +1,315 @@
+
+LN_LAMBDA_BMAX_FACTOR = 3 # BMAX fitting factor for ln_Lambda, 3 is the default value meaning b_max = r/3
+LN_LAMBDA_BMIN_FACTOR = 3 # BMIN fitting factor for ln_Lambda, 3 is the default value meaning b_min = dc/3 (dc is the spatial resolution in simulation)
+
```
done

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Tools/A_SET_IC.py:
```diff
@@ -0,0 +1,382 @@
+
+PURE_TABLE = params["PURE_TABLE"]
+if PURE_TABLE: pure_table_path = params["Chandrasekhar_DF_Table_Name"]
+else : pure_table_path = ""
```
changed to read from `log` file and `Record_*`

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Tools/A_SET_IC.py:
```diff
@@ -0,0 +1,382 @@
+    if Stellar == (0, 0, 0, "None"):
+        print("!!!!!!!!REMINDER : No stellar component!!!!!!!!")
+
+    halo_cutoff_mass_ = accumulated_mass_from_file(GC_ri)
```
That is just for testing with an without underscore case, `A_SET_IC.py` is deleted now

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/README.md:
```diff
@@ -0,0 +1 @@
+Please visit https://hackmd.io/@Q-7np4JqRmmhXMvo0h5Fqg/B1L_61Yb6 to see the details
```
We have tested the case with AMR on, but this didn't match the chandrasekhar's result, so we decided not to implement AMR to avoid further confusion

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Input__TestProb_Quick_Burkert:
```diff
@@ -0,0 +1,58 @@
+STELLAR_Rt            0                             #*[pc][0]
+
+# (2) Total Cloud Parameters
+Cloud_CenterX         10E3                          #*central x coordinate in [pc] SHOULD BE SET BY USER [38E3]
```
deleted `[pc]` and all other units in `Input__*`

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Input__TestProb_Quick_Burkert:
```diff
@@ -0,0 +1,58 @@
+
+# (2) Total Cloud Parameters
+Cloud_CenterX         10E3                          #*central x coordinate in [pc] SHOULD BE SET BY USER [38E3]
+Cloud_CenterY         10E3                          #*central y coordinate in [pc] SHOULD BE SET BY USER [38E3]
```
- changed to `e`
- I though it is more intuitive to write in `e3` cause it can be convert to `kpc` by just canceling the `e3`

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Tools/C_Chandrasekhar_dynamical_friction.py:
```diff
@@ -0,0 +1,315 @@
+date = datetime.datetime.now().strftime("%Y%m%d")
+
+if WRITE_TO_FILE:
+    filename = f'Chandrasekhar_result/bminFactor={LN_LAMBDA_BMIN_FACTOR}_bmaxFactor={LN_LAMBDA_BMAX_FACTOR}_{date}.txt'
```
fixed

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Tools/A_SET_IC.py:
```diff
@@ -0,0 +1,382 @@
+def getrho(mass, r0, rt, model_name):
+    if model_name == "None":
+        return 0
+    rho_0 = 3.0
```
done

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Input__TestProb_Quick_Burkert:
```diff
@@ -0,0 +1,58 @@
+Cloud_RSeed           123                           #!random seed for setting particle position and velocity (>=0) [123]
+
+PURE_TABLE            0                             #* Determine whether using (1) to create the table or using user specified table to calculate initial condition [0]
+Chandrasekhar_DF_Table_Name   Burkert_table_20240102.txt #+ if PURE_TABLE = 0; *if PURE_TABLE=1 the table must contains radius, density and enclosed mass
```
Yes, this is deleted and changed to `Density_Table_Name`

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Tools/A_SET_IC.py:
```diff
@@ -0,0 +1,382 @@
+            return 4*np.pi*x*x*(r0**3) * smooth_transition_factor_* King_dens(x)
+        else:
+            return 0
+    f = integrate.nquad(massbase, [[0, x]])[0] * rho_s
```
done with `f` and checked other variables

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Tools/A_SET_IC.py:
```diff
@@ -0,0 +1,382 @@
+
+    def massbase(x):
+
+        smooth_transition_factor_ = smooth_transition(x*r0,begin_smooth_r, end_smooth_r, GC_ri)
```
deleted `A_SET_IC.py` and there is no more trailing underscore in the variables

ej-1211 commented:
@hsinhaoHHuang (@hyschive) Sorry for the late update, I have chosen the **1st suggestion** above and have already fixed most of the problems (comments and suggestions) above, besides the fixing, in this version, we
- Add the function of subtracting GC's potential so that it will be more accurate to find the center 
- Three cases (NFW, Burkert, DoublePowerLaw) are tested again and results are inside `README`
- deleted the unnecessary script (`python` and `.cpp`) and unsupported feature

Please review my pull request again and please tell me if you have any suggestion! 
Thank you very much !!!!

hsinhaoHHuang commented on src/TestProblem/Hydro/DynamicalFriction/Init_TestProb_Hydro_DynamicalFriction.cpp:
```diff
@@ -0,0 +1,448 @@
+
+
+// C. Set the center finding method base on Input__TestProb
+   if ( FixCenter )
+   {
+      if ( MPI_Rank == 0 )
+         {
+         char Filename[MAX_STRING];
+         sprintf( Filename, "%s", "Record__Result_Center" );
+         FILE *File = fopen( Filename, "a" );
+         if ( Time[0]==0.0 )
+         {
+            fprintf(File, "%15s\t%15s\t%15s\t%15s\n", "#          Time", " CenterX", " CenterY", " CenterZ");
+         }
+         fprintf(File, "%15.7f\t%15.7e\t%15.7e\t%15.7e\n",Time[0]  ,amr->BoxCenter[0], amr->BoxCenter[1], amr->BoxCenter[2]);
+         fclose ( File );
+      }
+   }
```
When `FixCenter` is on, it seems it just keeps recording the box centers, which are all constants.
It seems unnecessary and wasting time to me.
How about *always* recording the minimum potential in `Record__Result_Center`?
When we need to fix the center in the post-analysis, we can easily ignore `Record__Result_Center` and use the box center as the center instead.
Then, I think we don't need the parameter `FixCenter` in the simulations.

hsinhaoHHuang commented on src/TestProblem/Hydro/DynamicalFriction/Init_TestProb_Hydro_DynamicalFriction.cpp:
```diff
@@ -0,0 +1,448 @@
+   {
+      Aux_Message( stdout, "=============================================================================\n" );
+      Aux_Message( stdout, "  test problem ID     = %d\n",     TESTPROB_ID );
+      Aux_Message( stdout, "  GC small Gas        = %13.5e\n", GC_SmallGas );
+      Aux_Message( stdout, "  GC initial radius   = %13.5e\n", GC_R );
+      Aux_Message( stdout, "  GC Mass             = %13.5e\n", GC_MASS );
+      Aux_Message( stdout, "  Fix Center          = %d\n",     FixCenter );
+      Aux_Message( stdout, "  Search Radius       = %13.5e\n",  SearchRadius );
+      Aux_Message( stdout, "  Pure Table          = %d\n",     PURE_TABLE );
+      Aux_Message( stdout, "  Density Table Name  = %s\n",     TableName );
+      Aux_Message( stdout, "  Halo Type           = %s\n",     HaloType );
+      Aux_Message( stdout, "  HALO_RHO_0          = %13.5e\n", Halo_Rho0 );
+      Aux_Message( stdout, "  HALO_Rs             = %13.5e\n", Halo_Rs );
+      Aux_Message( stdout, "  HALO_Rt             = %13.5e\n", Halo_Rt );
+      Aux_Message( stdout, "  Halo_Profle_Param_a = %13.5f\n", Halo_Profile_Param_a );
+      Aux_Message( stdout, "  Halo_Profle_Param_b = %13.5f\n", Halo_Profile_Param_b );
+      Aux_Message( stdout, "  Halo_Profle_Param_c = %13.5f\n", Halo_Profile_Param_c );
```
How about also using `%13.5e`?

hsinhaoHHuang replied:
How about using `%13.7e` for better alignment?

hsinhaoHHuang commented on src/TestProblem/Hydro/DynamicalFriction/Init_TestProb_Hydro_DynamicalFriction.cpp:
```diff
@@ -0,0 +1,448 @@
+      if ( MPI_Rank == 0 )
+         {
+         char Filename[MAX_STRING];
+         sprintf( Filename, "%s", "Record__Result_Center" );
```
How about specifying that it is the minimum potential in the filename?

hsinhaoHHuang commented on src/TestProblem/Hydro/DynamicalFriction/Init_TestProb_Hydro_DynamicalFriction.cpp:
```diff
@@ -0,0 +1,448 @@
+         {
+            fprintf(File, "%15s\t%15s\t%15s\t%15s\n", "#          Time", " CenterX", " CenterY", " CenterZ");
+         }
+         fprintf(File, "%15.7f\t%15.7e\t%15.7e\t%15.7e\n",
```
How about also printing the Time as `%15.7e`?

hsinhaoHHuang commented on src/TestProblem/Hydro/DynamicalFriction/Calculate_IC.cpp:
```diff
@@ -0,0 +1,269 @@
+								    const double Halo_Rho0, const double Halo_Rs, const double Halo_Rt,
+								    const char* Table_Name, const bool Pure_Table){
+
+   if ( MPI_Rank == 0 )    Aux_Message( stdout, "%s ...\n", __FUNCTION__ );
+   if ( MPI_Rank == 0 )    Aux_Message( stdout, "%s -> HaloType \n", HaloType );
+   if ( MPI_Rank == 0 )    Aux_Message( stdout, "%s -> Table_Name \n", Table_Name );
```
How about:
```
if ( MPI_Rank == 0 )    Aux_Message( stdout, "%s ...\n", __FUNCTION__ );
if ( MPI_Rank == 0 )    Aux_Message( stdout, "HaloTye    : %s\n", HaloType   );
if ( MPI_Rank == 0 )    Aux_Message( stdout, "Table_Name : %s\n", Table_Name );
```

ej-1211 replied:
Fixed

hsinhaoHHuang reviewed (changes requested):
Hi @ej-1211, Thanks for your update!

I have finished my second review and left some comments.
Please take another look, and please let me know if I didn't make myself clear.
Thanks!

Here is the result I got from the default set-up (`DoublePowerLaw`)
The vertical lines look a little weird to me but I am not sure whether this is expected.
![Compare_result](https://github.com/gamer-project/gamer/assets/24509818/fa73d575-59c5-4dde-a2fc-de25f5bd8f83)

ChunYen-Chen commented on example/test_problem/Hydro/DynamicalFriction/Input__Profile_Params:
```diff
@@ -0,0 +1,23 @@
+# Halo/Cloud Profile Parameter 
+# ===============================================================================================================
+Cloud_CenterX             48e3                          # central x coordinate in [48e3]
+Cloud_CenterY             48e3                          # central y coordinate in [48e3]
+Cloud_CenterZ             48e3                          # central z coordinate in [48e3]
```
Should it be `4.8e4`?

ej-1211 replied:
Changed to correct scientific notation.

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Input__Parameter:
```diff
@@ -0,0 +1,249 @@
+
+
+# simulation scale
+BOX_SIZE                      96e+03      # box size along the longest side (in Mpc/h if COMOVING is adopted)
```
Changed to the correct scientific notation.

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Input__TestProb:
```diff
@@ -0,0 +1,32 @@
+SEARCH_RADIUS              2                             # search radius for the potential minimum search (unit:spatial resolution)
+                                                         # If FIX_CENTER = 1, this parameter is useless
+# (1) HALO and STELLAR Parameters
+HALO_TYPE                  DoublePowerLaw                # the type of the halo (NFW,Burkert,King,DoublePowerLaw)[DoublePowerLaw]
```
@hsinhaoHHuang Do you mean in the comment area? If the answer is yes, I have fixed it!

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Input__TestProb:
```diff
@@ -0,0 +1,32 @@
+
+
+# (0) Simulation configuration
+FIX_CENTER	           0				 # 1:Fix the center to be the halo center 0:dynamically adjust it base on potential minimum [0]
```
Fixed, sorry for my poor grammar

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Input__TestProb:
```diff
@@ -0,0 +1,32 @@
+
+# (3) Inherit from ParticleEquilibrium (basically NO CHANGE NEEDED) 
+GC_SmallGas                1e-10                         # The maximum tolerance gas density for the simulation [1e-10]
+ParEqmIC_Params_Filenames  Input__Profile_Params         # The filenames of the cloud info (which wwe want to set the Equilibrium IC)
```
Fixed

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/Input__TestProb:
```diff
@@ -0,0 +1,32 @@
+# (3) Inherit from ParticleEquilibrium (basically NO CHANGE NEEDED) 
+GC_SmallGas                1e-10                         # The maximum tolerance gas density for the simulation [1e-10]
+ParEqmIC_Params_Filenames  Input__Profile_Params         # The filenames of the cloud info (which wwe want to set the Equilibrium IC)
+ParEqmIC_Cloud_Num         1                             # Ttotal number of cloud [1]
```
Fixed

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/README:
```diff
@@ -0,0 +1,92 @@
+   - only support one massive object (GC), and doesn't support MPI.
+   - This test problem implements some features in different test problems, which is `ParticleEquilibriumIC`
+2. Three default cases are provided in this test problem, see 2. Initial conditions. for the parameters.
+   - For the simulation result for the three cases, please refer to 
```
Fixed

ej-1211 commented on example/test_problem/Hydro/DynamicalFriction/README:
```diff
@@ -0,0 +1,92 @@
+2. Initial conditions:
+   (1) Burkert Density Profile :
+      (a) Input__Parameter :
+          - BOX_SIZE                  16e+03
```
Have changed to the correct scientific notation.

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Calculate_IC.cpp:
```diff
@@ -0,0 +1,269 @@
+      return rho_0 * King_dens(x);
+   } else if ( modelName == "DoublePowerLaw" )
+   {
+      return rho_0 * DoublePowerLaw_dens(x, Halo_Profile_Param_a,Halo_Profile_Param_b,Halo_Profile_Param_c);
```
Fixed

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Calculate_IC.cpp:
```diff
@@ -0,0 +1,269 @@
+   } else if ( r_value > end )
+   {
+      return exp(-(r_value - start) / scale);
+   } else {
```
Fixed

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Calculate_IC.cpp:
```diff
@@ -0,0 +1,269 @@
+      }
+      for (int i=0; i<nbin; ++i)
+      {
+         fprintf(File, "%15.7e\t%15.7e\t%15.7e\n", r[i],dens[i],mass[i]);
```
Fixed

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Calculate_IC.cpp:
```diff
@@ -0,0 +1,269 @@
+// 1. Find the index of radius
+   int index = -1;
+   double min_diff = abs(r[0] - GC_R);
+   for (int i = 1; i < nbin + 1; ++i)
```
Fixed

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Calculate_IC.cpp:
```diff
@@ -0,0 +1,269 @@
+
+   double v_initial = sqrt(NEWTON_G * interpolated_mass / GC_R);
+
+   vector<double> GC_position = { amr->BoxSize[0]*0.5 + GC_R, amr->BoxSize[0]*0.5, amr->BoxSize[0]*0.5};
+   vector<double> GC_velocity = { 0.0, v_initial * interpolated_mass / ( GC_MASS + interpolated_mass)   , 0.0};
+   vector<double> Halo_velocity = {0.0, -v_initial * GC_MASS / ( GC_MASS + interpolated_mass) , 0.0};
```
Aligned the equal signs

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Calculate_IC.cpp:
```diff
@@ -0,0 +1,269 @@
+      } else if ( modelName == "King" )
+      {
+         return prefactor * smooth_transition_factor * King_dens(x);
+      } else if ( modelName == "DoublePowerLaw" ) {
```
Fixed

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Calculate_IC.cpp:
```diff
@@ -0,0 +1,269 @@
+   }
+
+// Write the result to a file
+   if ( MPI_Rank == 0)
```
Fixed

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Calculate_IC.cpp:
```diff
@@ -0,0 +1,269 @@
+      char Filename[MAX_STRING];
+      sprintf( Filename, "%s", "Profile_Table" );
+      FILE *File = fopen( Filename, "a" );
+      if (Time[0]==0.0)
```
Fixed

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Calculate_IC.cpp:
Done with all functions

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Calculate_IC.cpp:
```diff
@@ -0,0 +1,269 @@
+{
+   double h = (b - a) / n; // Step size
+   double sum = 0.5 * (f(a) + f(b)); // Start with end points
+   for (int i = 1; i < n; i++)
```
Done

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Calculate_IC.cpp:
Done in `Calculate_IC.cpp`

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Calculate_IC.cpp:
```diff
@@ -0,0 +1,269 @@
+{
+   vector<double> result(n);
+   double delta = (end - start) / (n - 1);
+   for(int i = 0; i < n; ++i)
```
Done

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Calculate_IC.cpp:
Fixed

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Calculate_IC.cpp:
```diff
@@ -0,0 +1,269 @@
+      FILE *File = fopen( Filename, "a" );
+      if (Time[0]==0.0)
+      {
+         fprintf(File, "%15s\t%15s\t%15s\n", "Radius", "Density", "Enclosed Mass");
```
added

ej-1211 commented on src/TestProblem/Hydro/DynamicalFriction/Init_TestProb_Hydro_DynamicalFriction.cpp:
```diff
@@ -0,0 +1,448 @@
+   ReadPara->Add( "HALO_Rt",                 &Halo_Rt,               NoDef_double,  NoMin_double,     NoMax_double      );
+   ReadPara->Add( "Halo_Profile_Param_a",    &Halo_Profile_Param_a,  NoDef_double,  NoMin_double,     NoMax_double      );
+   ReadPara->Add( "Halo_Profile_Param_b",    &Halo_Profile_Param_b,  NoDef_double,  NoMin_double,     NoMax_double      );
+   ReadPara->Add( "Halo_Profile_Param_c",    &Halo_Profile_Param_c,  NoDef_double,  NoMin_double,     NoMax_double      );
```
Fixed

hyschive commented:
@ej-1211 Just a reminder. Let's get it done this summer :muscle:
